In [96]:
import pandas as pd #tablas https://pandas.pydata.org/docs/user_guide/pyarrow.html
import numpy as np #estadistica https://numpy.org/doc/1.25/user/index.html#user
import matplotlib.pyplot as plt #gráficos https://matplotlib.org/stable/plot_types/index.html
import seaborn as sns #gráficos https://seaborn.pydata.org/tutorial/regression.html
from sklearn import linear_model #modelo de regresion https://scikit-learn.org/stable/
from sklearn.preprocessing import MinMaxScaler # do not use the function Normalise() - it does something entirely different
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from datetime import datetime #https://docs.python.org/3/library/datetime.html
from datetime import timedelta #this is a good time to learn about dates and timestamps. You can find some info here:
from sklearn.linear_model import LogisticRegression
from sklearn.utils import resample
from imblearn.over_sampling import SMOTE
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.neighbors import KNeighborsClassifier
from scipy.stats import ttest_1samp
import scipy.stats as st
from sklearn.neighbors import KNeighborsClassifier
from mlxtend.plotting import plot_decision_regions
import pandas as pd
import numpy as np
import pymysql                        # for getting data from a SQL database
from sqlalchemy import create_engine  # for establishing the connection and authentication
from getpass import getpass  # To get the password without showing the input

In [97]:
exdt = pd.read_csv('expiration_date_KEY.csv')

In [98]:
exdt['expiration_date'] = pd.to_datetime(exdt['expiration_date'])

1. BRINGING ALL THE TWEETS 4M

In [99]:
password = getpass()

In [100]:
connection_string = 'mysql+pymysql://root:'+password+'@localhost/raw_data'
engine = create_engine(connection_string)
query = '''SELECT * FROM company_tweets_raw_2015_2020;'''

tweets = pd.read_sql_query(query, engine)

In [101]:
tweets

tweet_id company_id          writer  post_date  \
0         550803612197457920       AAPL      SentiQuant 2015-01-01   
1         550803610825928706       AAPL      SentiQuant 2015-01-01   
2         550803225113157632       AAPL     MacHashNews 2015-01-01   
3         550802957370159104       AAPL   WaltLightShed 2015-01-01   
4         550802855129382912       AAPL    2waystrading 2015-01-01   
...                      ...        ...             ...        ...   
3717959  1212158772015034369       TSLA       StockJoke 2019-12-31   
3717960  1212159099632267268       TSLA       sweetog76 2019-12-31   
3717961  1212159184931717120       TSLA  MelaynaLokosky 2019-12-31   
3717962  1212159838882533376       TSLA   ShortingIsFun 2019-12-31   
3717963  1212160015332728833       TSLA     Commuternyc 2019-12-31   

                                                      body  comment_num  \
0        #TOPTICKERTWEETS $AAPL $IMRS $BABA $EBAY $AMZN...            0   
1        #SENTISHIFTUP $K $FB $GOOGL $GS $GOLD $T $AAPL...            0   
2        Rumor Roundup: What to expect when you're expe...            0   
3        An $AAPL store line in Sapporo Japan for the "...            2   
4        $AAPL - Will $AAPL Give Second entry opportuni...            0   
...                                                    ...          ...   
3717959  I'm getting wasted on NYE, and thanks to @NHTS...            0   
3717960                       $TSLA Very Interesting dude:            0   
3717961  If only 2020 brought more self-awareness to Te...            0   
3717962  In 2020 I may start Tweeting out positive news...            0   
3717963  Patiently Waiting for the no twitter sitter tw...            0   

         retweet_num  like_num  
0                  0         1  
1                  0         1  
2                  0         0  
3                  4         4  
4                  0         0  
...              ...       ...  
3717959            0         0  
3717960            0         0  
3717961            1         0  
3717962            0         1  
3717963            0         5  

[3717964 rows x 8 columns]

In [102]:
aggregated_tweets= tweets.groupby(['post_date',"company_id"]).agg({'body': 'count', 'retweet_num': 'mean', 'like_num': 'mean', "comment_num":"mean", }).reset_index()

In [103]:
aggregated_tweets_sorted = aggregated_tweets.sort_values(by=['company_id', 'post_date'])
aggregated_tweets_sorted 


post_date company_id  body  retweet_num   like_num  comment_num
0    2015-01-01       AAPL   299     4.180602   1.331104     0.170569
5    2015-01-02       AAPL   867     2.115340   0.504037     0.117647
10   2015-01-03       AAPL   346     0.271676   0.468208     0.069364
15   2015-01-04       AAPL   406     0.812808   0.817734     0.135468
20   2015-01-05       AAPL  1147     0.360070   0.517001     0.126417
...         ...        ...   ...          ...        ...          ...
9102 2019-12-27       TSLA  1150     1.956522  11.660870     1.282609
9107 2019-12-28       TSLA   564     0.865248   7.916667     1.015957
9112 2019-12-29       TSLA   578     1.247405  10.204152     1.122837
9117 2019-12-30       TSLA  1261     1.706582  10.659794     1.197462
9122 2019-12-31       TSLA   918     1.053377   9.303922     0.896514

[9123 rows x 6 columns]

In [104]:
tweets_agg_exp = aggregated_tweets_sorted .merge(exdt, how='left', left_on='post_date', right_on='expiration_date')

In [105]:
tweets_agg_exp

post_date company_id  body  retweet_num   like_num  comment_num  \
0    2015-01-01       AAPL   299     4.180602   1.331104     0.170569   
1    2015-01-02       AAPL   867     2.115340   0.504037     0.117647   
2    2015-01-03       AAPL   346     0.271676   0.468208     0.069364   
3    2015-01-04       AAPL   406     0.812808   0.817734     0.135468   
4    2015-01-05       AAPL  1147     0.360070   0.517001     0.126417   
...         ...        ...   ...          ...        ...          ...   
9118 2019-12-27       TSLA  1150     1.956522  11.660870     1.282609   
9119 2019-12-28       TSLA   564     0.865248   7.916667     1.015957   
9120 2019-12-29       TSLA   578     1.247405  10.204152     1.122837   
9121 2019-12-30       TSLA  1261     1.706582  10.659794     1.197462   
9122 2019-12-31       TSLA   918     1.053377   9.303922     0.896514   

     expiration_date  
0                NaT  
1                NaT  
2                NaT  
3                NaT  
4                NaT  
...              ...  
9118             NaT  
9119             NaT  
9120             NaT  
9121             NaT  
9122             NaT  

[9123 rows x 7 columns]

In [106]:
indexes = tweets_agg_exp [tweets_agg_exp ['expiration_date'].notnull()].index
indexes_0 = np.insert(indexes, 0, 0)

In [107]:
indexes_0

Int64Index([   0,   15,   49,   77,  104,  131,  166,  193,  228,  255,
            ...
            8803, 8831, 8894, 8929, 8957, 8985, 9020, 9048, 9076, 9111],
           dtype='int64', length=296)

In [108]:
def convert_to_ranges(index_array):
    ranges = []
    start_index = index_array[0]
    for i in range(1, len(index_array)):
        ranges.append((start_index, index_array[i]+1))
        start_index = index_array[i]+1
    ranges.append((start_index, index_array[-1]))
    return ranges

index_ranges = convert_to_ranges(indexes_0) 
index_ranges 

[(0, 16),
 (16, 50),
 (50, 78),
 (78, 105),
 (105, 132),
 (132, 167),
 (167, 194),
 (194, 229),
 (229, 256),
 (256, 283),
 (283, 318),
 (318, 345),
 (345, 373),
 (373, 408),
 (408, 436),
 (436, 464),
 (464, 499),
 (499, 527),
 (527, 555),
 (555, 590),
 (590, 618),
 (618, 653),
 (653, 681),
 (681, 709),
 (709, 744),
 (744, 772),
 (772, 800),
 (800, 835),
 (835, 863),
 (863, 891),
 (891, 926),
 (926, 954),
 (954, 982),
 (982, 1017),
 (1017, 1045),
 (1045, 1073),
 (1073, 1108),
 (1108, 1136),
 (1136, 1164),
 (1164, 1199),
 (1199, 1227),
 (1227, 1255),
 (1255, 1290),
 (1290, 1318),
 (1318, 1353),
 (1353, 1381),
 (1381, 1409),
 (1409, 1444),
 (1444, 1472),
 (1472, 1500),
 (1500, 1528),
 (1528, 1591),
 (1591, 1626),
 (1626, 1654),
 (1654, 1682),
 (1682, 1717),
 (1717, 1745),
 (1745, 1773),
 (1773, 1808),
 (1808, 1835),
 (1835, 1870),
 (1870, 1898),
 (1898, 1926),
 (1926, 1954),
 (1954, 1989),
 (1989, 2017),
 (2017, 2052),
 (2052, 2080),
 (2080, 2108),
 (2108, 2143),
 (2143, 2171),
 (2171, 21

In [109]:
tweet_count = [tweets_agg_exp.loc[start:end, 'body'].sum() for start, end in index_ranges]
tweet_count = tweet_count[:-1]

In [110]:
len(tweet_count)

295

In [111]:
retweet_mean_per = [tweets_agg_exp.loc[start:end, 'retweet_num'].mean() for start, end in index_ranges]

In [112]:
likes_mean_per = [tweets_agg_exp.loc[start:end, 'like_num'].mean() for start, end in index_ranges]

In [113]:
comment_mean_per = [tweets_agg_exp.loc[start:end, 'comment_num'].mean() for start, end in index_ranges]

In [114]:
retweet_mean_per = retweet_mean_per[:-1]
likes_mean_per = likes_mean_per[:-1]
comment_mean_per = comment_mean_per[:-1]

In [115]:
len(retweet_mean_per )

295

In [116]:
tweets_agg = tweets_agg_exp.dropna()

In [117]:
columns_to_drop = ["body", "retweet_num", "like_num", "comment_num", "post_date"]
tweets_agg = tweets_agg.drop(columns=columns_to_drop)

In [120]:
tweets_agg = tweets_agg .reset_index(drop=True)

In [121]:
tweets_agg["tweet_count"] = tweet_count
tweets_agg["likes_mean"] = likes_mean_per
tweets_agg["retweet_mean"] = retweet_mean_per
tweets_agg["comment_mean"] = comment_mean_per

In [122]:
tweets_agg

company_id expiration_date  tweet_count  likes_mean  retweet_mean  \
0         AAPL      2015-01-16        14508    0.552529      0.664074   
1         AAPL      2015-02-20        43674    0.580656      0.569494   
2         AAPL      2015-03-20        42708    0.500023      0.500350   
3         AAPL      2015-04-17        26685    0.449074      0.439766   
4         AAPL      2015-05-15        32051    0.466478      0.415809   
..         ...             ...          ...         ...           ...   
290       TSLA      2019-08-16        19737   10.545790      1.647369   
291       TSLA      2019-09-20        21123   11.558816      1.888907   
292       TSLA      2019-10-18        20516   10.896159      1.477335   
293       TSLA      2019-11-15        25569   11.606386      1.695980   
294       TSLA      2019-12-20        31628   11.887997      1.838464   

     comment_mean  
0        0.118066  
1        0.139502  
2        0.118108  
3        0.097524  
4        0.181444  
..            ...  
290      1.085467  
291      1.119495  
292      1.075981  
293      1.169661  
294      1.211627  

[295 rows x 6 columns]

In [124]:
tweets_agg["company_id"].unique()

array(['AAPL', 'AMZN', 'GOOG', 'MSFT', 'TSLA'], dtype=object)

AND THAT'S HOW 4M TWEETS ARE FILTERED INTO 295 ROWS

In [123]:
connection_string = 'mysql+pymysql://root:'+password+'@localhost/twitter_stock_all'
engine = create_engine(connection_string)

In [126]:
#tweets_agg.to_sql('tweets_agg_2015_2020', engine, if_exists='append', index=False)